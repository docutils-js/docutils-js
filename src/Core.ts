import {Publisher} from './Publisher';
import * as defaults from './defaults';
import { ConfigSettings, SettingsSpecType } from './types';
import Writer from './Writer';
import Parser from './Parser';
import Reader from './Reader';
import { Settings } from '../gen/Settings';
export { Publisher };

export const defaultUsage = '%%prog [options] [<source> [<destination>]]';
export const defaultDescription = ('Reads from <source> (default is stdin) and writes to <destination> (default is stdout).  See <http://docutils.sf.net/docs/user/config.html> for the full reference.');

/* We need a non command-line parsing based function */

/* We should document all the arguments to this function */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function publish(args: any): void {
    const _defaults = {
        readerName: defaults.defaultReaderName,
        parserName: defaults.defaultParserName,
        usage: defaultUsage,
        description: defaultDescription,
        enableExitStatus: defaults.defaultEnableExitStatus,
    };
    const myArgs = { ..._defaults, ...args };
    const {
        reader, readerName, parser, parserName, writer, writerName,
        settings,
    } = myArgs;
    const pub = new Publisher({
        reader, parser, writer, settings,
    });
    pub.setComponents(readerName, parserName, writerName);
    //pub.publish2({
    //    argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
    //});
}


export interface PublishCmdLineArgs {
reader?: Reader;
readerName?: string;
parser?: Parser;
parserName?: string;
writer?: Writer;
writerName?: string;
settings?: Settings;
settingsSpec?: SettingsSpecType[];
settingsOverrides?: ConfigSettings;
configSection?: string;
enableExitStatus?: boolean;
argv?: string[];
usage?: string,
description?: string;
}

/**
    Set up & run a `Publisher` for command-line-based file I/O (input and
    output file paths taken automatically from the command line).  Return the
    encoded string output also.

    Parameters: see `publish_programmatically` for the remainder.

    - `argv`: Command-line argument list to use instead of ``sys.argv[1:]``.
    - `usage`: Usage string, output if there's a problem parsing the command
      line.
    - `description`: Program description, output for the "--help" option
      (along with command-line option descriptions).
*/
/// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function publishCmdLine(args: PublishCmdLineArgs, cb: any): void {
    const _defaults = {
        readerName: 'standalone',
        parserName: 'restructuredtext',
        usage: defaultUsage,
        description: defaultDescription,
        enableExitStatus: true,
    };
    args = { ..._defaults, ...args };
    const {
        reader, readerName, parser, parserName, writer, writerName,
        settings, settingsSpec, settingsOverrides, configSection,
        enableExitStatus, argv, usage, description,
    } = args;
    const pub = new Publisher({
        reader, parser, writer, settings,
    });
    pub.setComponents(readerName, parserName, writerName);
    pub.publish({
        argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
    }, cb);
}
