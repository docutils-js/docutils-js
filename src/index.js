export const __version__ = '0.14js';
