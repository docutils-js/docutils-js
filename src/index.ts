/* eslint-disable-next-line import/prefer-default-export */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

export const __version__ = '0.14js';

export {
    nodes, Writer, Transform,
};
