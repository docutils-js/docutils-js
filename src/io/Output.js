import TransformSpec  from '../TransformSpec';

class Output extends TransformSpec {
    constructor(destination, destinationPath, encoding, errorHandler) {
	super({});
	this.encoding = encoding;
	this.errorHandler = errorHandler || 'strict'
	this.destination = destination;
	this.destinationPath = destinationPath;
	if(!destinationPath) {
	    this.destinationPath = this.defaultDestinationPath;
	}
    }

    write(data) {
	process.stdout.write(data);
    }

    encode(data) {
    }

    toString() {
	return `Output<${this.constructor.name}>`;
    }

}

export default Output;
