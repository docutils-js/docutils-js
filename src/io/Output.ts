import TransformSpec from '../TransformSpec';

class Output<T> extends TransformSpec {
    public componentType: string = 'output';
    public supported: string[] = [];
    protected defaultDestinationPath: any;
    private destinationPath: any;
    private encoding: any;
    public destination?: T;
    private errorHandler: string;
    constructor(destination?: T, destinationPath?: any, encoding?: any, errorHandler?: any) {
        super();
        this.encoding = encoding;
        this.errorHandler = errorHandler || 'strict';
        this.destination = destination;
        this.destinationPath = destinationPath;
        if (!destinationPath) {
            this.destinationPath = this.defaultDestinationPath;
        }
    }

    /* istanbul ignore method */
    /* eslint-disable-next-line no-unused-vars */
    write(data: string): void {
    }

    /* istanbul ignore method */
    encode(data: string): string {
        return data; // fixme?
    }

    /* istanbul ignore method */
    toString() {
        return `Output<${this.constructor.name}>`;
    }
}
//Output.componentType = 'Output';
//Output.defaultDestinationPath = null;

export default Output;
