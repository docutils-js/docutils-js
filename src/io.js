import TransformSpec  from './TransformSpec';
import { createReadStream } from 'fs';
import ErrorOutput from './ErrorOutput';
import Input from './io/Input';
import Output from './io/Output';
import { ApplicationError } from'./Exceptions';

export class StringInput extends Input {
    constructor(...args) {
	super(...args);
	this.sourcePath = '<string>';
    }
	
    read(cb) {
	cb(undefined, this.source);
    }
}

export class StringOutput extends Output {
    constructor(...args) {
	super(...args);
	this.defaultDestinationPath = '<string>'
    }
    
    write(data) {
        //self.destination = self.encode(data) // fixme encoding
	if(Array.isArray(data)) {
	    data = JSON.stringify(data);
	}
	this.destination = data;
	return this.destination;
    }
}

export class FileInput extends Input {
    /* ew, too much logic for a constructor, with side effects etc! */
    constructor(args) {
	super(args);
	let { source, sourcePath, encoding, errorHandler, autoClose,
	      mode } = args;
	if(typeof source === 'undefined' && typeof sourcePath === 'undefined') {
	    throw new ApplicationError("FileInput: Undefined source and sourcePath");
	}
	
	if(autoClose === undefined) {
	    autoClose = true;
	}
	if(mode === undefined) {
	    mode = 'r';
	}
	
	this.autoClose = autoClose;
	this._stderr = new ErrorOutput()
	if(!source) {
	    if(sourcePath) {
		try {
		    this.source = createReadStream(sourcePath, { encoding: 'utf8' });
		}
		catch(error) {
		    console.log(error.stack);
		    throw error;
		}
	    } else {
		this.source = process.stdin;
	    }
	} else {
	    
	    // ??
	}
	if(!sourcePath) {
	    this.sourcePath = this.source.name;
	}
    }

    /* Read and decode a single file and return the data (Unicode string).
     */
    read(cb) {
	setTimeout(() => {
	let data;
	try {
	    /* reading ? */
	    if(this.source === process.stdin) {
		// do stuff
	    } else {
		data = this.source.read();
		if(data === null) {
		    console.log("read returned null ?");
		}
		cb(undefined, data);
	    }
	} catch(error) {
	    console.log(error.stack);
	    cb(error, undefined);
	}
	}, 100);
    }

    readLines() {
	return this.read().splitlines(true);
    }

    close() {
	if(this.source !== process.stdin) {
	    this.source.close();
	}
    }
}

export class FileOutput extends Output {
    
}
