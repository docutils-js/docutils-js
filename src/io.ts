import Input from './io/Input';
import Output from './io/Output';
import { ReadInputCallback, LoggerType } from "./types";

/** Direct string input. */
export class StringInput extends Input {
    public constructor(source: string, logger: LoggerType, sourcePath?: string, encoding?: string, errorHandler?: string) {
        super({source, sourcePath, encoding, errorHandler, logger});
        this.sourcePath = '<string>';
    }

    public read(cb: ReadInputCallback<string|string[]|{}>): void {
        cb(undefined, this.source);
    }
}

export class StringOutput extends Output<string> {
    public constructor(
        logger: LoggerType,
        destination?: string,
        destinationPath?: string,
        encoding?: string,
        errorHandler?: string
    ) {
        super(logger, destination, destinationPath, encoding, errorHandler);
        this.defaultDestinationPath = '<string>';

    }

    public write(data: string): string {
        // self.destination = self.encode(data) // fixme encoding
        if (Array.isArray(data)) {
            data = JSON.stringify(data);
        }
        this.destination = data;
        return this.destination;
    }
}

