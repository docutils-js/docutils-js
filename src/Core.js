import readers from './Readers';
import writers from './Writers';
import { OptionParser } from './FrontEnd';
import { Source } from './Sources';
import { FileInput, FileOutput } from './io';
import Publisher from './Publisher';

export { Publisher };

export const defaultUsage = '%prog [options] [<source> [<destination>]]';
export const defaultDescription = ('Reads from <source> (default is stdin) and writes to <destination> (default is stdout).  See <http://docutils.sf.net/docs/user/config.html> for the full reference.');


export function publishCmdLine(args, cb) {
    const _defaults = {
 readerName: 'standalone',
                        parserName: 'restructuredtext',
                        usage: defaultUsage,
                        description: defaultDescription,
                        enableExitStatus: true,
};
    args = { ..._defaults, ...args };
    const {
 reader, readerName, parser, parserName, writer, writerName,
      settings, settingsSpec, settingsOverrides, configSection,
            enableExitStatus, argv, usage, description,
} = args;
    const pub = new Publisher({
 reader, parser, writer, settings,
});
    pub.setComponents(readerName, parserName, writerName);
    pub.publish({
argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus,
}, cb);
}
