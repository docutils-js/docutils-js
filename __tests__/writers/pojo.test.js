import { Publisher, publishCmdLine, defaultDescription } from '../../src/Core';
import { Source } from '../../src/Sources';
import { StringInput, StringOutput } from '../../src/io';
import Output from '../../src/io/Output';
jest.mock('../../src/io/Output');

const currentLogLines = [];

afterEach(() => {
    if(currentLogLines.length) {
	console.log(currentLogLines.join('\n') + '\n');
	currentLogLines.length = 0;
    }
});

const defaultArgs = {
    readerName: 'standalone',
    parserName: 'restructuredtext',
    usage: '',
    description: '',
    enableExitStatus: true,
    writerName: 'pojo',
};

/* should pull from baseSettings */
const defaultSettings = {
    debug: true,
    autoIdPrefix: 'auto',
    idPrefix: '',
};

test.only('rst2pojo pipeline', () => {
    const settings = { ...defaultSettings };
    const args = { ...defaultArgs };

    const debugLog = [];
    const debugFn = (msg) => {
	console.log(msg);
//	currentLogLines.push(msg);
    };

    const { readerName, parserName, writerName } = args;
    const source = new StringInput({ source: `Random test
===========
I like food.

`});
    
    const destination = new StringOutput({});
    const pub = new Publisher({
	source, destination, settings, debug: true, debugFn
    });
    pub.setComponents(readerName, parserName, writerName);
    return new Promise((resolve, reject) => {
	pub.publish({}, (error, ...args) => {
	    if (error) {
		reject(error);
		return;
	    }
	    expect(JSON.parse(destination.destination)).toMatchSnapshot();
	    currentLogLines.length = 0;
	    resolve();
	});
    });
});

