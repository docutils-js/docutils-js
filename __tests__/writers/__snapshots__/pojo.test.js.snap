// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rst2pojo pipeline 1`] = `
Array [
  "document",
  Object {
    "source": "<string>",
  },
  Array [
    Array [
      "section",
      Object {
        "ids": Array [
          "random test",
        ],
        "names": Array [
          "random test",
        ],
      },
      Array [
        Array [
          "title",
          Object {},
          Array [
            "Random test",
          ],
        ],
        Array [
          "paragraph",
          Object {},
          Array [
            "I like food.",
          ],
        ],
      ],
    ],
  ],
]
`;
